estrest/src/mapper/event_structure_mapper.py

```diff
@@ -0,0 +1,82 @@
+from typing import Set
+from itertools import combinations
+
+from causality import CausalModel
+from causality.causal_model import VALS
+from event import Event
+from event_structure import EventStructure
+from utils import powerset, ids
+
+
+class EventStructureToCausalModelMapper:
+    def __init__(self, es: EventStructure):
+        self.es = es
+
+    @staticmethod
+    def __conflict_var(e: Event, ep: Event):
+        ids_ = ids({e, ep})
+        return f'C{tuple(sorted(ids_))}'.replace('\'', '')
+
+    def __add_conflict_vars(self, cm: CausalModel):
+        for e, ep in combinations(self.es.events, 2):
+            var = self.__conflict_var(e, ep)
+            if self.es.conflicts(e, ep):
+                cm.add_constant(var, True)
+            else:
+                cm.add_constant(var, False)
+
+    def con(self, s: Set[Event], vals: VALS):
+        result = True
+        for e, ep in combinations(s, 2):
+            result = result and not vals[self.__conflict_var(e, ep)]
+        return result
+
+    @staticmethod
+    def __min_en_var(s: Set[Event], e: Event):
+        ids_ = ids(s)
+        return f'M{sorted(ids_), e}'.replace('\'', '')
+
+    def min(self, s: Set[Event], e: Event, vals: VALS):
+        result = True
+        for sp in powerset(self.es.events):
+            if (sp < s or s < sp) and (e not in sp):
+                result = result and not vals[self.__min_en_var(sp, e)]
+        return result
+
+    def __add_min_en_vars(self, cm: CausalModel):
+        for s in powerset(self.es.events):
+            for e in self.es.events:
+                if e in s:
+                    continue
+                var = self.__min_en_var(s, e)
+                if self.es.min_en(s, e):
+                    cm.add(var, lambda vals, s_=s, e_=e: self.min(s_, e_, vals) and self.con(s_, vals))
```

Comment:
The method you pass to add must only accept a single `vals` arguemnt.
Furthermore, here you do not need to declare them as arguments since you have access to them.

Reply:
Solution:

We define an outer lambda that binds `s_` and `e_`, then in that lambda, we write the lambda that only takes `vals`, then we call the outer lambda; so that in the end, `s_` and `e_` are bound and we have a lambda that takes only `vals`:

```
(lambda s_=s, e_=e: lambda vals: ...)()
```